Migrator looks for work too frequently

Based on access logs and REST find requests seen in graphite it looks like migrator looks for work every 1 or 2 seconds.  Frequent is OK, but it shouldn't be this frequent.

Guess I was tired last night or just missed it.  The migrator is not able to parse the response for getting the next available job.  Not sure why yet.  The response looks good at a glance, it's valid JSON.  

```
Exception in thread "ConsistencyCheckerRunner" com.netflix.hystrix.exception.HystrixRuntimeException: migrator:primaryClient failed and no fallback available.
	at com.netflix.hystrix.AbstractCommand$16.call(AbstractCommand.java:760)
	at com.netflix.hystrix.AbstractCommand$16.call(AbstractCommand.java:747)
	at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$1.onError(OperatorOnErrorResumeNextViaFunction.java:77)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at com.netflix.hystrix.AbstractCommand$DeprecatedOnFallbackHookApplication$1.onError(AbstractCommand.java:1388)
	at com.netflix.hystrix.AbstractCommand$FallbackHookApplication$1.onError(AbstractCommand.java:1278)
	at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:314)
	at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:306)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7466)
	at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$1.onError(OperatorOnErrorResumeNextViaFunction.java:78)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at com.netflix.hystrix.AbstractCommand$HystrixObservableTimeoutOperator$3.onError(AbstractCommand.java:921)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:70)
	at com.netflix.hystrix.AbstractCommand$DeprecatedOnRunHookApplication$1.onError(AbstractCommand.java:1347)
	at com.netflix.hystrix.AbstractCommand$ExecutionHookApplication$1.onError(AbstractCommand.java:1247)
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:297)
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:289)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7466)
	at com.netflix.hystrix.AbstractCommand$1.call(AbstractCommand.java:386)
	at com.netflix.hystrix.AbstractCommand$1.call(AbstractCommand.java:366)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7466)
	at com.netflix.hystrix.AbstractCommand$ObservableCommand$1.call(AbstractCommand.java:1019)
	at com.netflix.hystrix.AbstractCommand$ObservableCommand$1.call(AbstractCommand.java:1015)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.subscribe(Observable.java:7556)
	at rx.internal.operators.BlockingOperatorToFuture.toFuture(BlockingOperatorToFuture.java:57)
	at rx.observables.BlockingObservable.toFuture(BlockingObservable.java:410)
	at com.netflix.hystrix.HystrixCommand.queue(HystrixCommand.java:379)
	at com.netflix.hystrix.HystrixCommand.execute(HystrixCommand.java:335)
	at com.redhat.lightblue.client.hystrix.LightblueHystrixClient.data(LightblueHystrixClient.java:95)
	at com.redhat.lightblue.migrator.consistency.ConsistencyChecker.getNextAvailableJob(ConsistencyChecker.java:225)
	at com.redhat.lightblue.migrator.consistency.ConsistencyChecker.run(ConsistencyChecker.java:161)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.redhat.lightblue.client.http.LightblueHttpClientException: Error sending lightblue request: {"query":{"field":"whenAvailableDate","op":">=","rvalue":"20150318T14:08:36.951+0000"},"projection":[{"field":"*","include":true,"recursive":true}],"sort":[{"whenAvailableDate":"$asc"}], "range": [0,1]}
	at com.redhat.lightblue.client.http.LightblueHttpClient.data(LightblueHttpClient.java:136)
	at com.redhat.lightblue.client.hystrix.LightblueHystrixClient$DataTypeHystrixCommand.run(LightblueHystrixClient.java:69)
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:294)
	... 48 more
Caused by: com.redhat.lightblue.client.response.LightblueResponseParseException: Error parsing lightblue response: {"status":"COMPLETE","modifiedCount":0,"matchCount":6927,"processed":[{"startDate":"20150318T14:00:00.000+0000","jobExecutions":[{"actualEndDate":"20150317T22:20:14.988+0000","ownerName":"migrator","consistentDocumentCount":0,"completedFlag":true,"hostName":"hostname","actualStartDate":"20150317T22:20:11.653+0000","overwrittenDocumentCount":0,"pid":"2484@hostname","inconsistentDocumentCount":0,"processedDocumentCount":0}],"endDate":"20150318T15:00:00.000+0000","whenAvailableDate":"20150318T15:00:00.000+0000","configurationName":"entityName","lastUpdateDate":"20150317T13:51:23.000+0000","creationDate":"20150317T13:51:23.000+0000","expectedExecutionMilliseconds":30000,"jobExecutions#":1,"createdBy":"nmalik","_id":"entityNameJob_203848","lastUpdatedBy":"nmalik","objectType":"migrationJob"},{"startDate":"20150318T15:00:00.000+0000","jobExecutions":[{"actualEndDate":"20150317T22:20:15.250+0000","ownerName":"migrator","consistentDocumentCount":0,"completedFlag":true,"hostName":"hostname","actualStartDate":"20150317T22:20:11.935+0000","overwrittenDocumentCount":0,"pid":"2484@hostname","inconsistentDocumentCount":0,"processedDocumentCount":0}],"endDate":"20150318T16:00:00.000+0000","whenAvailableDate":"20150318T16:00:00.000+0000","configurationName":"entityName","lastUpdateDate":"20150317T13:51:23.000+0000","creationDate":"20150317T13:51:23.000+0000","expectedExecutionMilliseconds":30000,"jobExecutions#":1,"createdBy":"nmalik","_id":"entityNameJob_203849","lastUpdatedBy":"nmalik","objectType":"migrationJob"}]}

	at com.redhat.lightblue.client.response.LightblueResponse.parseProcessed(LightblueResponse.java:119)
	at com.redhat.lightblue.client.http.LightblueHttpClient.data(LightblueHttpClient.java:134)
	... 50 more
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of com.redhat.lightblue.migrator.consistency.MigrationJob out of START_ARRAY token
 at [Source: N/A; line: -1, column: -1]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:164)
	at com.fasterxml.jackson.databind.DeserializationContext.mappingException(DeserializationContext.java:575)
	at com.fasterxml.jackson.databind.DeserializationContext.mappingException(DeserializationContext.java:569)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromArray(BeanDeserializerBase.java:1121)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:123)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:2860)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1569)
	at com.redhat.lightblue.client.response.LightblueResponse.parseProcessed(LightblueResponse.java:117)
	... 51 more
```

--- a/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
+++ b/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
@@ -151,6 +151,8 @@ public class ConsistencyChecker implements Runnable {
                     }
                 }
             }
+            
+            LOGGER.info("Done setting up job executors");
 
             if (executors.isEmpty()) {
                 if (run && !Thread.interrupted()) {
@@ -183,10 +185,13 @@ public class ConsistencyChecker implements Runnable {
                     }
                 }
             }
+            
+            LOGGER.info("Job executors done");
         }
     }
 
     protected List<MigrationJob> getMigrationJobs(MigrationConfiguration configuration) {
+        LOGGER.info("Loading jobs for {}", configuration.getConfigurationName());
         List<MigrationJob> jobs = new ArrayList<>();
         try {
             DataFindRequest findRequest = new DataFindRequest("migrationJob", migrationJobEntityVersion);

--- a/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
+++ b/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
@@ -192,7 +192,7 @@ public class ConsistencyChecker implements Runnable {
             DataFindRequest findRequest = new DataFindRequest("migrationJob", migrationJobEntityVersion);
             findRequest.where(and(
                     withValue("configurationName = " + configuration.getConfigurationName()),
-                    withValue("whenAvailableDate >= " + ClientConstants.getDateFormat().format(new Date())),
+                    withValue("whenAvailableDate <= " + ClientConstants.getDateFormat().format(new Date())),
                     not(withSubfield("jobExecutions", withValue("completedFlag = true")))));
             findRequest.select(includeFieldRecursively("*"));
             jobs.addAll(Arrays.asList(client.data(findRequest, MigrationJob[].class)));

--- a/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
+++ b/consistency-checker/src/main/java/com/redhat/lightblue/migrator/consistency/ConsistencyChecker.java
@@ -33,9 +33,9 @@ public class ConsistencyChecker implements Runnable {
 
     public static final Logger LOGGER = LoggerFactory.getLogger(ConsistencyChecker.class);
 
-    public static final int MAX_JOB_WAIT_TIME_MSEC = 24 * 60 * 60 * 1000; // 24 hours
-    public static final int MAX_THREAD_WAIT_TIME_MSEC = 6 * 60 * 60 * 1000; // 6 hours
-    public static final int DEFAULT_WAIT = 30000;         // 30 minutes
+    public static final int MAX_JOB_WAIT_MSEC = 24 * 60 * 60 * 1000; // 24 hours
+    public static final int MAX_THREAD_WAIT_MSEC = 6 * 60 * 60 * 1000; // 6 hours
+    public static final int DEFAULT_JOB_WAIT_MSEC = 30 * 60 * 1000; // 30 minutes
 
     private String consistencyCheckerName;
     private String hostName;
@@ -155,12 +155,16 @@ public class ConsistencyChecker implements Runnable {
             if (executors.isEmpty()) {
                 if (run && !Thread.interrupted()) {
                     MigrationJob nextJob = getNextAvailableJob();
-                    long timeUntilNextJob = DEFAULT_WAIT;
-                    if (nextJob != null) {
+                    long timeUntilNextJob = DEFAULT_JOB_WAIT_MSEC;
+                    if (nextJob != null && nextJob.getWhenAvailableDate() != null) {
                         timeUntilNextJob = nextJob.getWhenAvailableDate().getTime() - new Date().getTime();
                     }
+                    if (timeUntilNextJob > MAX_JOB_WAIT_MSEC) {
+                        timeUntilNextJob = MAX_JOB_WAIT_MSEC;
+                    }
                     try {
-                        Thread.sleep((timeUntilNextJob > MAX_JOB_WAIT_TIME_MSEC) ? MAX_JOB_WAIT_TIME_MSEC : timeUntilNextJob);
+                        LOGGER.info("Waiting for next job, sleeping for {} msec", timeUntilNextJob);
+                        Thread.sleep(timeUntilNextJob);
                     } catch (InterruptedException e) {
                         run = false;
                     }
@@ -172,7 +176,7 @@ public class ConsistencyChecker implements Runnable {
                 if ((!Thread.interrupted())) {
                     try {
                         for (ExecutorService executor : executors) {
-                            executor.awaitTermination(MAX_THREAD_WAIT_TIME_MSEC, TimeUnit.MILLISECONDS);
+                            executor.awaitTermination(MAX_THREAD_WAIT_MSEC, TimeUnit.MILLISECONDS);
                         }
                     } catch (InterruptedException e) {
                         run = false;
@@ -188,6 +192,7 @@ public class ConsistencyChecker implements Runnable {
             DataFindRequest findRequest = new DataFindRequest("migrationJob", migrationJobEntityVersion);
             findRequest.where(and(
                     withValue("configurationName = " + configuration.getConfigurationName()),
+                    withValue("whenAvailableDate >= " + ClientConstants.getDateFormat().format(new Date())),
                     not(withSubfield("jobExecutions", withValue("completedFlag = true")))));
             findRequest.select(includeFieldRecursively("*"));
             jobs.addAll(Arrays.asList(client.data(findRequest, MigrationJob[].class)));
@@ -210,13 +215,17 @@ public class ConsistencyChecker implements Runnable {
         return configurations;
     }
 
+    /**
+     * Gets the next job available for processing.
+     * @return 
+     */
     protected MigrationJob getNextAvailableJob() {
         MigrationJob job = null;
         try {
             DataFindRequest findRequest = new DataFindRequest("migrationJob", migrationJobEntityVersion);
             findRequest.where(withValue("whenAvailableDate >= " + ClientConstants.getDateFormat().format(new Date())));
             findRequest.sort(new SortCondition("whenAvailableDate", SortDirection.ASC));
-            findRequest.range(0, 1);
+            findRequest.range(0, 0); // range is inclusive
             findRequest.select(includeFieldRecursively("*"));
             job = client.data(findRequest, MigrationJob.class);
         } catch (IOException e) {

